File: plep_main/src/deltadak/commands/DeleteCommand.java
package deltadak.commands;

import deltadak.Database;
import deltadak.database.DatabaseFacade;
import deltadak.ui.AbstractController;
import deltadak.HomeworkTask;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeView;

import javax.xml.crypto.Data;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

/**
 * Command to delete an {@link HomeworkTask} from a TreeView.
 */
public class DeleteCommand extends Command {

    /** remember state */
    protected AbstractController controller; // program to an interface so tests can provide a dummy
    protected LocalDate dayState;
    protected List<List<HomeworkTask>> treeViewItems; // A list of (lists containing a parent and it's children)
    protected TreeView<HomeworkTask> tree;
    protected int indexState;
    protected List<HomeworkTask> deletedItemsList; // It's a list, includes subtasks

    /**
     * Constructor, gets selected index to find out what to delete.
     * @param controller the controller which needs to be updated after deletion
     * @param day from which to delete the task
     * @param treeViewItems corresponding to the day, items of listview, needed as parameter for testing (which can't use a listview)
     * @param index index of task in treeViewItems to delete, needed as parameter for testing (which can't use a listview)
     * @param tree to provide immediate user feedback
     */
    public DeleteCommand(AbstractController controller, LocalDate day, List<List<HomeworkTask>> treeViewItems, int index, TreeView<HomeworkTask> tree) {
        this.controller = controller;
        this.dayState = day;
        this.treeViewItems = new ArrayList<>(treeViewItems);
        this.indexState = index;
        this.tree = tree;
    }

    @Override
    protected void executionHook() {
        if (treeViewItems.isEmpty()) {
            throw new IllegalStateException("cannot delete item from empty treeview");
        }
        // save the deleted tasks so we can undo later
        deletedItemsList = treeViewItems.get(indexState);
        treeViewItems.remove(indexState);

        // use the treeview to delete an item so the user gets feedback
        if (tree != null) {

            // get the selected item BEFORE deleting the item, otherwise
            // we're selecting a different item
            TreeItem<HomeworkTask> selected = tree.getSelectionModel()
                    .getSelectedItem();

            // Delete the item from the 'expanded' table, which contains information whether the item was expanded or not.
//            controller.deleteExpanded(selected.getValue().getDatabaseID());

            tree.getRoot().getChildren().remove(indexState);
        }

        // use the items of the listview to update the database
//        controller.updateDatabase(dayState, treeViewItems);
//        System.out.println("id: " + deletedItemsList.get(0).getDatabaseID());
        Database.INSTANCE.deleteByID(deletedItemsList.get(0).getDatabaseID());
        controller.cleanUp(tree);
    }

    @Override
    protected void undoHook() {
        treeViewItems.add(indexState, deletedItemsList);
        // if not testing, provide user feedback
        if (tree != null) {
            // add parent task
            TreeItem<HomeworkTask> parent = new TreeItem<>(deletedItemsList.get(0));
            tree.getRoot().getChildren().add(indexState, parent);
            // add subtasks
            // start at i=1 because first item is the parent task
            for (int i = 1; i < deletedItemsList.size(); i++) {
                parent.getChildren().add(new TreeItem<>(deletedItemsList.get(i)));
            }

            new DatabaseFacade(controller.getProgressIndicator()).updateDatabase(dayState, treeViewItems);

//            int parentID = parent.getValue().getDatabaseID();
//            controller.insertExpandedItem(parentID, false);

            controller.cleanUp(tree);
        }
    }

    /**
     * Get the treeViewItems this command remembers, mainly used for testing purposes.
     * @return treeViewItems of {@link HomeworkTask}s.
     */
    public List<List<HomeworkTask>> getListItems() {
        return treeViewItems;
    }

}

File: src/main/kotlin/nl/deltadak/plep/commands/DeleteCommand.kt
package nl.deltadak.plep.commands

import javafx.scene.control.ProgressIndicator
import javafx.scene.control.TreeItem
import javafx.scene.control.TreeView
import nl.deltadak.plep.Database
import nl.deltadak.plep.HomeworkTask
import nl.deltadak.plep.database.DatabaseFacade
import nl.deltadak.plep.ui.treeview.TreeViewCleaner
import java.time.LocalDate

/**
 * Command to delete a [HomeworkTask] from a TreeView.
 */
open class DeleteCommand(
        /** To show progress. */
        val progressIndicator: ProgressIndicator,
        /** Day of the task to delete. */
        val day: LocalDate,
        /** All the tasks of that day. */
        treeViewItemsImmutable: List<List<HomeworkTask>>,
        /** Index of the task in treeViewItems. */
        private val index: Int,
        /** To provide feedback immediately in the TreeView. */
        val tree: TreeView<HomeworkTask>) : Command() {

    /** List including subtasks. */
    private var deletedItemsList: List<HomeworkTask> = emptyList()

    /** All the tasks of the day, making the given list mutable in order to edit, original reference need not be preserved. */
    var treeViewItems: MutableList<MutableList<HomeworkTask>> = treeViewItemsImmutable.toMutableList().map{it.toMutableList()}.toMutableList()

    override fun executionHook() {

        if (treeViewItems.isEmpty()) {
            throw IllegalStateException("cannot delete item from empty treeview")
        }

        // Save the deleted tasks so we can undo later.
        deletedItemsList = treeViewItems[this.index]
        treeViewItems.removeAt(this.index)

        // Known to be null when testing.
        if (tree.root != null) {
            // Use the TreeView to delete an item, to provide user feedback.
            tree.root.children.removeAt(this.index)

            Database.INSTANCE.deleteByID(deletedItemsList[0].databaseID)
            TreeViewCleaner().cleanSingleTreeView(tree)
        }

    }

    override fun undoHook() {
        treeViewItems.add(this.index, deletedItemsList.toMutableList())

        if (tree.root != null) {
            // First add parent.
            val parent = TreeItem<HomeworkTask>(deletedItemsList[0])
            tree.root.children.add(this.index, parent)

            // Then the children.
            deletedItemsList.subList(1, deletedItemsList.size)
                    .forEach {
                        parent.children.add(TreeItem<HomeworkTask>(it))
                    }

            DatabaseFacade(progressIndicator).pushData(day, treeViewItems)
            TreeViewCleaner().cleanSingleTreeView(tree)
        }

    }

}